# Aprendizaje automático. Práctica 2
## Marta Venegas Pardo

## Entrega 2.1

### Simple Linear Regression- Regresión lineal simple

#### Paso 1: Instalación de librerías

In [3]:
## Instalación de librerías
import numpy as np 
import pandas as pd
# import sklearn
# Modelo de regresión lineal
from sklearn.linear_model import LinearRegression
# Gráficas
import matplotlib.pyplot as plt

In [4]:
# LinearRegression: fit_intercept, normalize

#### Paso 2: Carga de datos

In [5]:
x = np.array([5, 15, 25, 35, 45, 55]).reshape((-1, 1)) # Variable explicativa
y = np.array([5, 20, 14, 32, 22, 38]) # Variable objetivo/respuesta

In [6]:
print('Regresor,x:\n',x,'\nDimensiones: ',x.shape,'\n\nRespuesta,y:\n',y,'\nDimensiones:',y.shape)

Regresor,x:
 [[ 5]
 [15]
 [25]
 [35]
 [45]
 [55]] 
Dimensiones:  (6, 1) 

Respuesta,y:
 [ 5 20 14 32 22 38] 
Dimensiones: (6,)


#### Paso 3: Creación del modelo y ajuste

In [7]:
# Definición del modelo
modelo = LinearRegression() # No añadimos hiperparámetros (fit_intercept: t independiente beta0, normalize: normalizar datos, n_jobs)

In [8]:
# Ajuste del modelo
modelo.fit(x,y) 

LinearRegression()

Con el argumento *.fit* estamos calculando los la estimaación de los coeficientes de regresión $\beta_0$ y $\beta_1$, obteniendo así $\hat\beta_0$ y $\hat\beta_1$ . 

Es decir, la ecuación de regresión linear simple es la siguiente:

<center>$ Y = \beta_0 + \beta_1 x + \epsilon$</center>

Y en el caso de la estimación:

<center>$\hat y = \hat\beta_0 + \hat\beta_1 x$</center>

In [9]:
# Otra forma más rápida:
model = LinearRegression().fit(x, y)

#### Paso 4: Obtención de los resultados

In [10]:
r_sq = model.score(x, y)
print(f"Coeficiente de determinación: {r_sq}")

Coeficiente de determinación: 0.715875613747954


Este coeficiente es una medida de la bondad de ajuste del modelo, en otras palabras, se trata de una medida de la bondad de ajuste de la recta a los datos. Se tiene que $R^2  = 0.716$, es decir, el modelo (la recta) explica casi un $72\%$ de la varaibilidad total de la variable respuesta.

Los coeficientes del modelo son los siguientes:

In [11]:
print(f"intercept: {modelo.intercept_}") # Intercept es un número

print(f"slope: {modelo.coef_}") # Slope es un array

intercept: 5.633333333333329
slope: [0.54]


Y por tanto, la ecuación de la recta de regresión es la siguiente: 
<center>$\hat y = 5,63+0,54x$</center>

Se tiene que: 

- El valor de $\beta_0 = 5,63$ indica que la varaible respuesta es 5,63 cuando $x=0$
- $\beta_0 = 0,54$ significa que el valor de y aumenta 0,54 cuando x aumenta en una unidad.

Ahora vamos a introducir $y$ como un array bidimensional, al igual que $x$, obteniendo así el mismo modelo:

In [12]:
Modelo2 = LinearRegression().fit(x, y.reshape((-1, 1)))
print(f"intercept: {Modelo2.intercept_}")

print(f"slope: {Modelo2.coef_}")

# Son ambos arrays (unidimensional y bidimensional, respectivamente)

intercept: [5.63333333]
slope: [[0.54]]


#### Paso 5: Cálculo de predicciones

In [13]:
y_pred = modelo.predict(x)
print(f"Predicciones:\n{y_pred}") # Array unidimensional

Predicciones:
[ 8.33333333 13.73333333 19.13333333 24.53333333 29.93333333 35.33333333]


Otra manera de realizar predicciones es sustituyendo los valores en la recta de regresión calculada, como vemos a continuación.

In [14]:
# Otra forma, utilizando beta0 y beta1

y_pred = modelo.intercept_ + modelo.coef_ * x
print(f"Respuesta:\n{y_pred}") # Array bidimensional 

Respuesta:
[[ 8.33333333]
 [13.73333333]
 [19.13333333]
 [24.53333333]
 [29.93333333]
 [35.33333333]]


In [17]:
# Otra forma, utilizando beta0 y beta1

y_pred = modelo.intercept_ + modelo.coef_ *  x.flatten()
print(f"Respuesta:\n{y_pred}") # Array bidimensional 

Respuesta:
[ 8.33333333 13.73333333 19.13333333 24.53333333 29.93333333 35.33333333]


Vamos ahora a utilizar el modelo en datos nuevos para así predecir el valor de Y para cada valor de x.

In [49]:
x_new = np.arange(5).reshape((-1, 1))
x_new

array([[0],
       [1],
       [2],
       [3],
       [4]])

Los valores de Y para estos valores de X son:

In [50]:
y_new = model.predict(x_new)
y_new

array([5.63333333, 6.17333333, 6.71333333, 7.25333333, 7.79333333])

In [51]:
x_new = x_new.ravel().tolist()
lista = list(zip(x_new,y_new))

In [60]:
print('x','\t','y_pred')
for i in lista:
    print(i)

x 	 y_pred
(0, 5.633333333333329)
(1, 6.173333333333329)
(2, 6.713333333333329)
(3, 7.25333333333333)
(4, 7.7933333333333294)


### Multiple Linear Regression - Regresión lineal múltiple

In [ ]:
# Las librerías ya están cargadas

## Entrega 2.2

### Polynomial Regression With scikit-learn


In [ ]:
# Librerías

from sklearn.preprocessing import PolynomialFeatures

### Advanced Linear Regression With statsmodels

In [ ]:
# Librerías
import statsmodels.api as sm